pilot/pkg/config/coredatamodel/discovery.go

```diff
@@ -0,0 +1,333 @@
+
+package coredatamodel
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+
+	"istio.io/api/annotation"
+	networking "istio.io/api/networking/v1alpha3"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/config/host"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/protocol"
+	"istio.io/pkg/log"
+)
+
+var (
+	notReadyEndpointkey                        = annotation.AlphaNetworkingNotReadyEndpoints.Name
+	_                   model.Controller       = &MCPDiscovery{}
+	_                   model.ServiceDiscovery = &MCPDiscovery{}
+)
+
+// DiscoveryOptions stores the configurable attributes of a Control
+type DiscoveryOptions struct {
+	XDSUpdater   model.XDSUpdater
+	Env          *model.Environment
+	ClusterID    string
+	DomainSuffix string
+}
+
+// mixerEnabled checks to see if mixer is enabled in the environment
+// so we can set the UID on eds endpoints
+func (o *DiscoveryOptions) mixerEnabled() bool {
+	return o.Env != nil && o.Env.Mesh != nil && (o.Env.Mesh.MixerCheckServer != "" || o.Env.Mesh.MixerReportServer != "")
+}
+
+// MCPDiscovery provides storage for NotReadyEndpoints
+type MCPDiscovery struct {
+	notReadyEndpointsMu sync.RWMutex
+	// [ip:port]config
+	notReadyEndpoints map[string]*model.Config
+	*DiscoveryOptions
+}
+
+// NewMCPDiscovery provides a new instance of Discovery
+func NewMCPDiscovery(options *DiscoveryOptions) *MCPDiscovery {
+	return &MCPDiscovery{
+		notReadyEndpoints: make(map[string]*model.Config),
+		DiscoveryOptions:  options,
+	}
+}
+
+// RegisterNotReadyEndpoints registers newly received NotReadyEndpoints
+// via MCP annotations
+func (d *MCPDiscovery) RegisterNotReadyEndpoints(conf *model.Config) {
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+
+	if nrEps, ok := conf.Annotations[notReadyEndpointkey]; ok {
+		addrs := strings.Split(nrEps, ",")
+		for _, addr := range addrs {
+			d.notReadyEndpoints[addr] = conf
+		}
+	}
+}
+
+// GetProxyServiceInstances returns service instances co-located with a given proxy
+func (d *MCPDiscovery) GetProxyServiceInstances(proxy *model.Proxy) ([]*model.ServiceInstance, error) {
+	out := make([]*model.ServiceInstance, 0)
+
+	// There is only one IP for kube registry
+	proxyIP := proxy.IPAddresses[0]
+
+	d.notReadyEndpointsMu.Lock()
+	defer d.notReadyEndpointsMu.Unlock()
+	for addr, conf := range d.notReadyEndpoints {
+		notReadyIP, port, err := net.SplitHostPort(addr)
+		if err != nil {
+			continue
+		}
+		notReadyPort, err := strconv.Atoi(port)
+		if proxyIP != notReadyIP || err != nil {
+			continue
+		}
+		se, ok := conf.Spec.(*networking.ServiceEntry)
+		if !ok {
+			return nil, errors.New("getProxyServiceInstances: wrong type")
+		}
+
+		for _, h := range se.Hosts {
+			for _, svcPort := range se.Ports {
+				// add not ready endpoint
+				out = append(out, d.notReadyServiceInstance(se, svcPort, conf, h, notReadyIP, notReadyPort))
```

Comment:
Thisis not right, for example, if 192.168.1.1:8080 and 192.168.1.1:8081 are not ready. What's the  output

Reply:
Can you elaborate a bit more? I have a test case for this case specifically. The expected out put would be 
```
 // NetworkEndpoint(endpoint{Address:192.168.1.1, Port:8080, servicePort: &{http-port 80 http}} service:{Hostname: svc.example2.com})
 // NetworkEndpoint(endpoint{Address:192.168.1.1, Port:8081, servicePort: &{http-port 80 http}} service:{Hostname: svc.example2.com})
```
is that not correct?